### ISSUE TITLE
Why is warning "4190" (-Wreturn-type-c-linkage) disabled on Windows?

### ISSUE THREAD

**Issue by marvin2k** [orocos-toolchain/rtt#125]
Two related questions I have after reading the compiler output.

- Why do the [Macros](https://github.com/orocos-toolchain/rtt/blob/master/rtt/plugin/ServicePlugin.hpp#L73) in ServicePlugin.hpp have C-linkage at all, when they return a C++ type?
- And why is the resulting [warning](https://github.com/orocos-toolchain/rtt/blob/master/rtt/plugin/ServicePlugin.hpp#L58) disabled for Windows builds?

Clang-3.5 on Linux will complain relatively often on using the macro:

> ...Plugin.cpp:51:1: warning: 'getRTTPluginName' has C-linkage specified, but returns user-defined type 'std::string' (aka 'basic_string<char, char_traits<char>, allocator<char> >') which is incompatible with C [-Wreturn-type-c-linkage]

I don't have an actual problem with that, it just looks not tidy ,-)

**Comment by meyerj** [comment]
I agree, that looks somehow wrong. `getRTTPluginName()` and `getRTTTargetName()` could simply return a `const char *` instead. Without the C-linkage declaration the generated symbol name required to lookup these functions in the PluginLoader would be implementation-defined. Probably the warning has been disabled for Windows builds because Visual C++ always had a problem with returning C++ types from a function with C-linkage, while in general it is not forbidden explicitly by the standard.

Unfortunately it would be hard to change the return type because a lot of plugin implementations do not use the macros defined in [ServicePlugin.hpp](https://github.com/orocos-toolchain/rtt/blob/master/rtt/plugin/ServicePlugin.hpp). There was already a discussion about this issue back in 2013 on the Orocos-users mailing list, including a possible solution:
http://www.orocos.org/forum/orocos/orocos-users/warning-getrttpluginname-has-c-linkage-specified-returns-user-defined-type

**Comment by marvin2k** [comment]
Plugins which are not using the API correctly prevent RTT from improving its codebase? And the fix is to silence the compiler warning. If you ask me: Break early, break loudly.

**Comment by meyerj** [comment]
I did not say that plugins do not use the API correctly. It's only that not all plugins do use the convenience macros, but the documentation clearly says "You *can* use this macro to make any Service available as a plugin.", not must. So changing the return type and the macros would break plugins that for some reason do not use it (e.g. in order to provide a global service). In any case they would have to be recompiled.
